### ISSUE TITLE
Have bot state accessors available from Middleware

### ISSUE THREAD

**Issue by MickH3** [Microsoft/botbuilder-dotnet#1218]
**State accessors only available from the bot, not middleware
Since all middleware runs as part of the 'turn', just like my bot code, I'd like to be able to get at bot state.  Currently, bot state is not available to me until my code hits the bot itself.  My bot's constructor gets called before the middleware gets called, so the correct bot state is available.  

**Describe the solution you'd like**
If I could get access to the TurnContext in the constructor of my bot, I could feed the accessor into TurnState to make it available from Middleware.

**Describe alternatives you've considered**
I am able to get at bot state from middleware by starting each middleware with await next()... first and my code follows.  It passes through all middleware and gets to my bot where I feed the accessors object I received in my constructor into the TurnState in ITurnContext.  

This is not ideal.  I'd much prefer to utilize the await next() mechanism in middleware for what it was designed for.

Thanks much,

Mick

[enhancement]

**Comment by JeffCordes** [comment]
Hello @MickH3 .  I am a fellow user..

If you create a constructor for your middleware that accepts the state accessors, you can pass the state accessors into it when you register the middleware to your bot in the startup.cs.

look at the auto save state middleware in this example..  https://github.com/Microsoft/BotBuilder-Samples/blob/master/samples/csharp_dotnetcore/51.cafe-bot/Startup.cs

**Comment by drub0y** [comment]
As @JeffCordes points out, you can create your property accessor and hand it to the middleware at construction time. That might look something like this:

#### Startup.cs

```CSharp
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        var storage = new MemoryStorage();
        var conversationState = new ConversationState(storage);

        // Create the property accessor here
        var myPropertyAccessor = conversationState.CreateProperty<MyPropertyValue>("MyProperty");

       // Register the accessor so it will be injected into the bot
       services.AddSingleton<IStatePropertyAccessor<MyPropertyValue>>(myPropertyAccessor);

        services.AddBot<MyBot>(options =>
        {
             // Explicitly pass the accessor to my middleware since middleware doesn't participate in DI
             options.Middleware.Add(new MyMiddleware(myPropertyAccessor));
        });
    }

    // remainder elided for brevity
}
```

**Comment by MickH3** [comment]
Thank you @JeffCordes  and @drub0y , I was able to get it worked out and all is well :).

I think the source of my confusion was an earlier version of this article:
https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-howto-v4-state?view=azure-bot-service-4.0&tabs=csharp

... which made it look like I had to declare my servcies.AddBot<> before I could add the Accessor Singleton.  It looks like it's been updated since.  The sample code there now reflects the same flexibility in order of operations you note above during ConfigureServices.

Thanks again!
